opendevin/controller/agent_controller.py

```diff
@@ -83,106 +76,79 @@ def __init__(
-        self.runtime.sandbox.close()
-        self.runtime.browser.close()
         await self.set_agent_state_to(AgentState.STOPPED)
 
-    def update_state_for_step(self, i):
-        self.state.iteration = i
-        self.state.background_commands_obs = self.runtime.get_background_obs()
+    def update_state_before_step(self):
+        self.state.iteration += 1
 
     def update_state_after_step(self):
         self.state.updated_info = []
 
-    async def add_error_to_history(self, message: str):
-        await self.add_history(NullAction(), ErrorObservation(message))
+    async def report_error(self, message: str):
+        await self.event_stream.add_event(ErrorObservation(message), EventSource.AGENT)
 
-    async def add_history(
-        self, action: Action, observation: Observation, add_to_stream=True
-    ):
-        if not isinstance(action, Action):
-            raise TypeError(
-                f'action must be an instance of Action, got {type(action).__name__} instead'
-            )
-        if not isinstance(observation, Observation):
-            raise TypeError(
-                f'observation must be an instance of Observation, got {type(observation).__name__} instead'
-            )
+    async def add_history(self, action: Action, observation: Observation):
         self.state.history.append((action, observation))
         self.state.updated_info.append((action, observation))
-        if add_to_stream:
-            await self.event_stream.add_event(action, EventSource.AGENT)
-            await self.event_stream.add_event(observation, EventSource.AGENT)
-
-    async def _run(self):
-        if self._agent_state != AgentState.RUNNING:
-            raise ValueError('Task is not in running state')
 
-        for i in range(self._cur_step, self.max_iterations):
-            self._cur_step = i
+    async def _start_step_loop(self):
+        while True:
             try:
-                finished = await self.step(i)
-                if finished:
-                    await self.set_agent_state_to(AgentState.FINISHED)
-                    break
-            except Exception:
-                logger.error('Error in loop', exc_info=True)
-                await self.set_agent_state_to(AgentState.ERROR)
-                await self.add_error_to_history(
-                    'Oops! Something went wrong while completing your task. You can check the logs for more info.'
-                )
+                await self._step()
+            except asyncio.CancelledError:
+                logger.info('AgentController task was cancelled')
                 break
-
-            if self._is_stuck():
-                logger.info('Loop detected, stopping task')
-                await self.set_agent_state_to(AgentState.ERROR)
-                await self.add_error_to_history(
-                    'I got stuck into a loop, the task has stopped.'
+            except Exception as e:
```

Comment:
Does this remove loop detection? It seems like this is somewhat important functionality

Reply:
Nope! You can see it still calls `_is_stuck`, it's just inside `_step` now